### ISSUE TITLE
Number field allows non numeric text entry

### ISSUE THREAD

**Issue by cuberoot** [adobe/spectrum-web-components#1514]
### Expected Behaviour

I would expect a number field to limit character entry to those that are valid in a number. React Spectrum seems to do a good job of this: https://react-spectrum.adobe.com/react-spectrum/NumberField.html

### Actual Behaviour

I can type in invalid characters such as 'abc'

#### Steps to Reproduce

1. browse to https://opensource.adobe.com/spectrum-web-components/storybook/?path=/story/number-field--default
2. click on the number field and enter 'abc'

![ScreenFlow](https://user-images.githubusercontent.com/154097/121140327-d25a2880-c88d-11eb-92df-e834af662a97.gif)

**Comment by Westbrook** [comment]
Can you think of any context where the keys here wouldn't be described as `\d` or all digits?

We overrode the `allowed-keys` when extending from `Textfield`, but it can be reapplied and the value can always be `\d`, or we could leave it open for consuming applications to customize/only set it to `\d` when not set. Does that seem like a proper approach to this situation?

**Comment by cuberoot** [comment]
Here's something interesting that I tried in the React spectrum field. I tried to enter `1.2.3` and it wouldn't let me put in the second `.`. So, it seems like it is checking if each keystroke results in something valid somehow. Also, we will need character input when there are units, right?

So, it seems like we need something more intelligent than just `/^\d+$/`? Thoughts?

**Comment by Westbrook** [comment]
Their implementation does not allow for inputting units.

We should be able to use the formatter to validate, and then update the value the same way that the `allowed-keys` functionality works in Textfield.

Is this something you'll need to be looking into for you team?

**Comment by cuberoot** [comment]
Yes, we have a lot of numeric inputs and would like to make sure that they're robust.

**Comment by bengfarrell** [comment]
I haven't looked over formatting yet on the number field, but in our designer's hopes and dreams, our internal app would do some of the more advanced things that some of our flagship products have done.

This includes entering 1+1 and getting a result of 2 on submitting. This also includes entering 1 inch into the numeric input and getting a result of 72px when hitting submit.

Just saw this issue when I was scrolling and wanted to offer these use cases which may or may not be blocked by this change.

**Comment by Westbrook** [comment]
Making some progress on an initial implementation of this: https://westbrook-number-field-validation--spectrum-web-components.netlify.app/components/number-field

@bengfarrell I do think that we will move into equation support in the future, and I'm not sure the two things here are blocking on each other, so I'm not binding them together at this time. Does your design have any thought on whether they'd expect "units" inside of their equations?

**Comment by Westbrook** [comment]
We're getting a little closer here! https://westbrook-number-field-validation--spectrum-web-components.netlify.app/storybook/?path=/story/number-field--default I _think_ we just need some good test coverage before we'll be ready to propose this as a PR. If you've got any super edge casey input you'd like included, please share it here!
